src/OSSupport/Network.cpp

```diff
@@ -0,0 +1,1259 @@
+
+	/** Mutex protecting all containers against multithreaded access. */
+	cCriticalSection m_CS;
+
+
+	/** Initializes the LibEvent internals. */
+	cNetworkSingleton(void);
+
+	/** Converts LibEvent-generated log events into log messages in MCS log. */
+	static void LogCallback(int a_Severity, const char * a_Msg);
+
+	/** Implements the thread that runs LibEvent's event dispatcher loop. */
+	static void RunEventLoop(cNetworkSingleton * a_Self);
+
+	/** Removes the specified hostname lookup from m_HostnameLookups.
+	Used by the underlying lookup implementation when the lookup is finished. */
+	void RemoveHostnameLookup(const cHostnameLookup * a_HostnameLookup);
+
+	/** Removes the specified IP lookup from m_IPLookups.
+	Used by the underlying lookup implementation when the lookup is finished. */
+	void RemoveIPLookup(const cIPLookup * a_IPLookup);
+
+	/** Removes the specified link from m_Connections.
+	Used by the underlying link implementation when the link is closed / errored. */
+	void RemoveLink(const cTCPLinkImpl * a_Link);
+};
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Globals:
+
+bool IsValidSocket(evutil_socket_t a_Socket)
+{
+	#ifdef _WIN32
+		return (a_Socket != INVALID_SOCKET);
+	#else  // _WIN32
+		return (a_Socket >= 0);
+	#endif  // else _WIN32
+}
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// cHostnameLookup:
+
+cHostnameLookup::cHostnameLookup(const AString & a_Hostname, cNetwork::cResolveNameCallbacksPtr a_Callbacks):
+	m_Callbacks(a_Callbacks),
+	m_Hostname(a_Hostname)
+{
+	evutil_addrinfo hints;
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_protocol = IPPROTO_TCP;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_flags = EVUTIL_AI_CANONNAME;
+	evdns_getaddrinfo(cNetworkSingleton::Get().m_DNSBase, a_Hostname.c_str(), nullptr, &hints, Callback, this);
+}
+
+
+
+
+
+void cHostnameLookup::Callback(int a_ErrCode, evutil_addrinfo * a_Addr, void * a_Self)
+{
+	// Get the Self class:
+	cHostnameLookup * Self = reinterpret_cast<cHostnameLookup *>(a_Self);
+	ASSERT(Self != nullptr);
+
+	// If an error has occurred, notify the error callback:
+	if (a_ErrCode != 0)
+	{
+		Self->m_Callbacks->OnError(a_ErrCode);
+		cNetworkSingleton::Get().RemoveHostnameLookup(Self);
+		return;
+	}
+
+	// Call the success handler for each entry received:
+	bool HasResolved = false;
+	evutil_addrinfo * OrigAddr = a_Addr;
+	for (;a_Addr != nullptr; a_Addr = a_Addr->ai_next)
+	{
+		char IP[128];
+		switch (a_Addr->ai_family)
+		{
+			case AF_INET:  // IPv4
+			{
+				sockaddr_in * sin = reinterpret_cast<sockaddr_in *>(a_Addr->ai_addr);
+				evutil_inet_ntop(AF_INET, &(sin->sin_addr), IP, sizeof(IP));
+				break;
+			}
+			case AF_INET6:  // IPv6
+			{
+				sockaddr_in6 * sin = reinterpret_cast<sockaddr_in6 *>(a_Addr->ai_addr);
+				evutil_inet_ntop(AF_INET6, &(sin->sin6_addr), IP, sizeof(IP));
+				break;
+			}
+			default:
```

Comment:
Not really. If a hostname resolves to IPv4, IPv6 and additionally, say, and IPX/SPX address, we want to ignore the IPX/SPX address silently.